File: Java Fundamentals/Session Five/Calculator.java
import java.util.*;
class Calculator{
    public static void main(String[] args){
        Scanner input = new Scanner(System.in);
        System.out.println("please enter the first number");
        double firstNumber = input.nextDouble();
        System.out.println("please enter the second number");
        double secondNumber = input.nextDouble();
        //check if user input an integer or double
        boolean firstN_isInteger = (firstNumber ==(int)firstNumber);
        boolean secondN_isInteger = (secondNumber ==(int)secondNumber);
        // let user choose an operation
        System.out.println("please enter the operator (+ , - , / , * , % )");
        char operator = input.next().charAt(0);
        switch (operator){
            case '+' :
                if(firstN_isInteger && secondN_isInteger){
                    sum((int)firstNumber , (int)secondNumber);
                }else {sum(firstNumber , secondNumber);}
                break;
            case '-' :
                if(firstN_isInteger && secondN_isInteger){
                    subtraction((int)firstNumber , (int)secondNumber);
                }else {subtraction(firstNumber , secondNumber);}
                break;
            case '*' :
                if(firstN_isInteger && secondN_isInteger){
                    multiplication((int)firstNumber , (int)secondNumber);
                }else {multiplication(firstNumber , secondNumber);}
                break;
            case '/' :
        if(firstN_isInteger && secondN_isInteger){
                    division((int)firstNumber , (int)secondNumber);
                }else {division(firstNumber , secondNumber);}
                break;
            case '%' :
                if(firstN_isInteger && secondN_isInteger){
                    reminder((int)firstNumber , (int)secondNumber);
                }else {reminder(firstNumber , secondNumber);}
                break;
            default :
             System.out.println("wrong input! ");
        }
    }
    public static void sum(int x, int y){System.out.println("The Summation is: " + (x + y) );}
    public static void sum(double x, double y){System.out.println("The Summation is: " + (x + y) );}
    public static void subtraction(int x, int y){System.out.println("The Subtraction is: " + (x - y) );}
    public static void subtraction(double x, double y){System.out.println("The Subtraction is: " + (x - y) );}
    public static void multiplication(int x, int y){System.out.println("The Multiplication is: " + (x * y) );}
    public static void multiplication(double x, double y){System.out.println("The Multiplication is: " + (x * y) );}
    public static void division(int x, int y){
        if(y==0)System.out.println("Error! division by zero is not allowed.");
        else System.out.println("The Division is: " + (x / y) );
    }
    public static void division(double x, double y){
         if(y==0)System.out.println("Error! division by zero is not allowed.");
        else System.out.println("The Division is: " + (x / y) );
    }
    public static void reminder(int x, int y){System.out.println("The Reminder is: " + (x + y) );}
    public static void reminder(double x, double y){System.out.println("The Reminder is: " + (x + y) );}
}

File: Kotlin Fundamentals/SessionOne/Calculator.kt
fun main(args: Array<String>){
    var stop :Boolean = true
    fun readNumbers():Double?{
        return try {
            readln()!!.toDouble()
        } catch (e: NumberFormatException) {
            stop = false
            return null
        }
    }
    fun readTxt():String?{
        return try {
            readln()
        } catch (error : NullPointerException) {
            println("Invalid input.")
            stop = false
            return null
        }
    }
loop@ do {
    print("Enter the first number: ")
    var firstNumb = readNumbers()
    print("\nEnter the second number: ")
    var secondNumb = readNumbers()
    println("Enter an operator: (  +  -  /  *  %  )" )
    var operator = readTxt()
//check input type and value
    if (firstNumb != null && secondNumb != null) {
        if (firstNumb.toInt().toDouble() == firstNumb) {
            firstNumb.toInt()
        }
        if (secondNumb.toInt().toDouble() == secondNumb){
            secondNumb.toInt()
        }
    }
    else if(operator == null || operator !in listOf("+", "-", "/", "*", "%")){
        println("You did not enter a valid operator!")
        stop = false
        continue
    }
    else{
        println("You have missed one of the two numbers or both of them!")
        stop = false
        continue
    }
//call suitable function
     when(operator){
        "+" -> summation (firstNumb,secondNumb)
        "-" -> subtract (firstNumb,secondNumb)
        "*" -> multiplication (firstNumb,secondNumb)
        "/" -> divided (firstNumb,secondNumb)
        "%" -> reminderM (firstNumb,secondNumb)
        else -> {
            println("Not an Operator! \n")
            stop = false
            continue@loop
        }
    }
    println("Do you want to make another calculation? (y/n)")
    var answer = readTxt()?.getOrNull(0)

    if (answer == null || answer.equals('n', ignoreCase = true)) {
        stop = true
    }
    else{
        stop=false
    }
}
while (!stop)
}

fun summation (numb1 : Int , numb2 : Int) {
    println("The summation = ${numb1+numb2}")
}
fun summation (numb1 : Double , numb2 : Double) {
    println("The summation = ${numb1+numb2}")
}
fun summation (numb1 : Double , numb2 : Int) {
    println("The summation = ${numb1+numb2}")
}
fun summation (numb1 : Int , numb2 : Double) {
    println("The summation = ${numb1+numb2}")
}
fun subtract (numb1 : Int , numb2 : Int) {
    println("The subtraction = ${numb1-numb2}")
}
fun subtract (numb1 : Double , numb2 : Double) {
    println("The subtraction = ${numb1-numb2}")
}
fun subtract (numb1 : Double , numb2 : Int) {
    println("The subtraction = ${numb1-numb2}")
}
fun subtract (numb1 : Int , numb2 : Double) {
    println("The subtraction = ${numb1-numb2}")
}
fun divided (numb1 : Int , numb2 : Int) {
    if (numb2==0){println("Error! Cannot divide by zero")
    } else {
        println("The division = ${numb1 / numb2}")
    }
}
fun divided (numb1 : Double , numb2 : Double) {
    if (numb2==0.0){println("Error! Cannot divide by zero")
    } else {
        println("The division = ${numb1 / numb2}")
    }
}
fun divided (numb1 : Int , numb2 : Double) {
    if (numb2==0.0){println("Error! Cannot divide by zero")
    } else println("The division = ${numb1/numb2}")
}
fun divided (numb1 : Double , numb2 : Int) {
    if (numb2==0){println("Error! Cannot divide by zero")
    } else println("The division = ${numb1/numb2}")
}
fun multiplication (numb1 : Int , numb2 : Int) {
    println("The multiplication = ${numb1*numb2}")
}
fun multiplication (numb1 : Double , numb2 : Double) {
    println("The multiplication = ${numb1*numb2}")
}
fun multiplication (numb1 : Double , numb2 : Int) {
    println("The multiplication = ${numb1*numb2}")
}
fun multiplication (numb1 : Int , numb2 : Double) {
    println("The multiplication = ${numb1*numb2}")
}
fun reminderM (numb1 : Int , numb2 : Int) {
    if (numb2==0){println("Error! Cannot divide by zero")
    } else {
        println("The reminder = ${numb1 % numb2}")
    }
}
fun reminderM (numb1 : Double , numb2 : Double) {
    if (numb2==0.0){println("Error! Cannot divide by zero")
    } else {
        println("The reminder = ${numb1 % numb2}")
    }
}
fun reminderM (numb1 : Double , numb2 : Int) {
    if (numb2==0){println("Error! Cannot divide by zero")
    } else {
        println("The reminder = ${numb1 % numb2}")
    }
}
fun reminderM (numb1 : Int , numb2 : Double) {
    if (numb2==0.0){println("Error! Cannot divide by zero")
    } else {
        println("The reminder = ${numb1 % numb2}")
    }
}